### DISCUSSION TITLE
Custom Shadow/Object Rendering

### DISCUSSION THREAD

**Discussion by fallenartist** [fabricjs/fabric.js#8584] [Q&A] [answered]
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have read and followed the [Issue Tracker Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md#%EF%B8%8F-issue-tracker)
- [X] I have searched and referenced existing issues and discussions
- [X] I am filing a **BUG** report.
- [X] I have managed to reproduce the bug **after** upgrading to the **latest** version
- [X] I have created an accurate and minimal reproduction

### Version

5.3.0

### In What environments are you experiencing the problem?

Chrome, Safari

### Node Version (if applicable)

None

### Link To Reproduction

https://jsfiddle.net/fallenartist/x27dsh35/25/

### Steps To Reproduce

1. Clone any shape multiple times with a shift in x/y position and as dense as necessary for the whole group's edge to create a smooth appearance.
2. Export the canvas with multiplier 4 (or final dimensions in range of 5000 px).
3. Observe smoothness of the resulting edges.

### Expected Behavior

I'm trying to achieve a 'long shadow' effect behind text and need the exported high resolution PNG to be as smooth as possible. To achieve it, I'm cloning the master text frame multiple times and with density geared towards the exported image size (print quality needed). I haven't found any other solution to achieve this programmatically but cloning the top object.

The longer the shadow the more expensive it gets because of the number of objects in question but my issue is mainly the quality of the resulting edge. No matter how dense I'm putting the cloned objects on canvas, I'm still seeing the jagged/stepped edge. Why is that and is there a way to make this better (also in terms of performance)?

Moreover, it looks like one side of the object is smoother than the other. See the attached pic of the exported canvas:

<img width="623" alt="fabric_js_canvas_export" src="https://user-images.githubusercontent.com/661820/211327719-3b3dff4c-d381-40e3-bed7-6fbe09dea058.png">

Side note: It would be great to have the 'long shadow' effect built in. Would you consider it? If not, did you come across any other example of this effect implemented for Fabric.js (I haven't found any). Thanks!

### Actual Behavior

The edges appear jagged/stepped no matter how dense the objects are put together.

**Comment by fallenartist** [comment]
I have used `cloneAsImage` here in addition to `clone` that looks better bar the slight blurriness outside the cloned text:

https://jsfiddle.net/fallenartist/x438swvq/

Is it possible to reduce that blurriness and also is there a way to merge/flatten all cloned layers into a single image?

**Comment by ShaMan123** [comment, marked as answer]
You are experiencing a canvas issue, pixel aliasing... not related to fabric it seems.
Making an image out it is better since it utilizes `imageSmoothing`
Looking at your example I must say it is an over kill.
I would try 2 approaches:
- Shadow with `nonScaling`, object set the the shadow size and transformed back to what is should look like
- create a custom object, that should be straight forward

---
- a bit more advanced is trying to use the cache canvas of the object to render the shadow but I wouldn't go there if I were you

**Reply by fallenartist** [reply]
Thanks, I suspected it being a limitation of canvas.

I used the only solution I could implement with my limited knowledge, so I'd appreciate if you could elaborate on the two proposed approaches (the custom object especially). A bit of code would help me to start tinkering. Thank you.

If I create a shadow object (why would I use `nonScaling`?) and clone it multiple times to create a long shadow as in my example, isn't it the same thing as `cloneAsImage` I used before performance wise?

Not getting anything [here](https://jsfiddle.net/fallenartist/x438swvq/62/) but surely I'm doing something wrong. Couldn't find anything referring to `.toObject()` method online:

```
function longShadow(el, length, angle, color) {
  const r = angle * (Math.PI / 180.0);
  const shadow = new fabric.Shadow({
    affectStroke: true,
    color: color,
    nonScaling: true
  });
  el.set({
    shadow: shadow
  });
  shadow.toObject(shadowobj => {
    shadowobj.set({
      left: el.left,
      top: el.top
    });
    for (let i = 1; i < length; i++) {
      shadowobj.clone(clone => {
        clone.set({
          left: el.left + (Math.cos(r) * i),
          top: el.top + (Math.sin(r) * i)
        });
      });
    }
  });
}
```

**Reply by ShaMan123** [reply]
https://jsfiddle.net/xp4k1rm5/

**Reply by ShaMan123** [reply]
with bbox
https://jsfiddle.net/xp4k1rm5/3/

**Reply by fallenartist** [reply]
Cool, thanks! One question – would it be possible to re-render the shadow after object rotation (or on the fly when transforming), so it maintains its angle and length relative to canvas? Thanks.

**Reply by ShaMan123** [reply]
You need to handle that
Turn off objectCaching

**Reply by fallenartist** [reply]
Thanks, will try. Can you just look at your bbox version and see if anything happens to controls frame when you change the angle to, say 135 deg?

**Reply by ShaMan123** [reply]
That case is a lot more complex since it alters left, top values so text positioning changes in relation to object's 0, 0.
Meaning that it will require a lot more adjustments

**Reply by ShaMan123** [reply]
In that case I would try 2 approaches:
1. adjust more text methods... perhaps adding an offset to all calculation somehow (maybe hacking it using a path property)
2. adjust the selection box rendering function as well as the function in charge determining if the cursor is inside the object (`calcOCoords` and or `_getTransformedDimensions`)

Highly experimental
Good luck

**Reply by ShaMan123** [reply]
I wonder if creating a font is easier or not...
But I guess you need to control the angle and stuff

**Reply by ShaMan123** [reply]
OK
So next idea would be to merge both of our efforts.
Take my rendering function but make the shadow a separate object and attach your listeners etc.
Then group the 2 together and the group will handle the bbox (if you use v6, if not regroup al the time)
Then the challenge would be interacting with the text which is handled by v6 but not v5

**Reply by fallenartist** [reply]
Thanks. It would be ideal for the shadow to be a separate, independent object to text, outside the text frame (imagine the shadow to be really long – it would be hard to even select the whole thing), not selectable on canvas etc. So if I understand it right, I'd use the extended `IText` in addition to regular editable `IText` and keep it in sync and recalculate on various transform events?

**Reply by fallenartist** [reply]
With what function do I adjust the object's 0, 0? I think it would be a matter of simply using the vector in `Math.cos(r) * this.longShadowSteps`. The width and height in `initDimensions()` should be right whatever the angle.

**Reply by ShaMan123** [reply]
It is not that simple IMO.
Since cursor logic isn't very flexible in terms of 0,0.
But why would you need to adjust it if using v6 Group? It handles all the headache for you.

**Reply by ShaMan123** [reply]
And if you must use v5 then simply connect the 2 objects and you are done
Modifying the rendering of the bbox will finalize your efforts

**Reply by fallenartist** [reply]
I can try v6 as I'm just starting anyway. But I still don't know how to modify the rendering, so it doesn't get clipped with pixels going to the negative of the origin.

**Reply by ShaMan123** [reply]
Since the shadow is a separate object it is a matter of positioning which is simple.
Handle it that way and the bbox will be flawless

**Reply by fallenartist** [reply]
Sorry, the separate shadow object would be the modified `IText` as in your example or something else entirely?

**Reply by ShaMan123** [reply]
Yes, the modified itext is the shadow object

**Reply by fallenartist** [reply]
OK, will play! Currently, the variables are hardcoded – do I add options to the object the same way as explained [here](http://fabricjs.com/fabric-intro-part-3#subclassing) (`type`, `initialize` and `toObject`)? Not that familiar with JS classes, I'm afraid.

I tried this but it looks like the options are not being inherited properly:

```
class Shadowed extends fabric.IText {

  //type: 'ShadowedIText'; //?

  initialize(options) {
    options || (options = {}); //not working?

    this.callSuper('initialize', options);
    this.shadowLength = options.shadowLength || 0;
    this.shadowAngle = options.shadowAngle || 90;
    this.shadowColor = options.shadowColor || 'rgba(0,0,0,1)';
  }

  toObject() {
    return fabric.util.object.extend(this.callSuper('toObject'), {
      shadowLength: this.get('shadowLength'),
      shadowAngle: this.get('shadowAngle'),
      shadowColor: this.get('shadowColor')
    });
  }
}
```

**Reply by ShaMan123** [reply]
yes
don't be
they don't bite usually

**Reply by fallenartist** [reply]
Working on it...

https://jsfiddle.net/fallenartist/6athd9r4/75/

`options || (options = {});` wrong syntax?

**Reply by ShaMan123** [reply]
https://jsfiddle.net/16xfhtsr/

**Reply by ShaMan123** [reply]
wrong constructor

**Reply by fallenartist** [reply]
Right. Not sure using a separate object would help? I'd need to render the shadow _after_ all transformations to text have been applied. Is 'baking' of object's transform possible?

Getting this strange artefacts when rotating: https://jsfiddle.net/fallenartist/6athd9r4/211/

**Reply by fallenartist** [reply]
It would seem the easiest option would be to enlarge the text object's canvas size (the area that gets redrawn with `dirty: true` flag, I guess?) evenly in all directions and set a reasonable maximum length of the shadow to fit within those bounds. Which setting is responsible for determining this size? Thanks for your assistance on this :)

<img width="543" alt="object-cache-canvas" src="https://user-images.githubusercontent.com/661820/212279140-847f12dc-da48-4604-b18f-55e0c0cdac93.png">

**Reply by ShaMan123** [reply]
First of all turn objectCaching off and make everything work as expected.
Then fine tune caching

**Reply by fallenartist** [reply]
These artefacts get drawn on the canvas with `objectCaching: false`. Not sure what you mean by fine tuning caching?

https://jsfiddle.net/fallenartist/6athd9r4/244/

**Reply by ShaMan123** [reply]
forcing no caching: https://jsfiddle.net/xr2ok8wd/

forcing bbox recalc (as pointer out you need to adjust position): https://jsfiddle.net/7rxe5awn/

**Reply by fallenartist** [reply]
Just a quickie as the code now fails and I couldn't pinpoint the breaking change ("TypeError: undefined is not an object (evaluating 'sel.set')"). What do I replace it with? Ta.

**Reply by ShaMan123** [reply]
It is your code...
simply no active selection and you try to use the controls to set the active selection...
Out of scope

**Reply by fallenartist** [reply]
Sorry! _Needs another coffee..._ 🫣

**Reply by fallenartist** [reply]
> https://jsfiddle.net/xr2ok8wd/

Tried to update this example to V6 beta but no luck. Can you help?  🙏

**Reply by ShaMan123** [reply]
FYI there is a pr I wrote fixing shadow projection rendering.
Search for it.
Do not use jsfiddle anymore.
Use the templates found in the issue template or via the sanbox cmd in our cli.
npm run sandbox start --help

**Comment by melchiar** [comment]
You could possibly resolve the jagged edges issue using the trick explained in the video below.

https://www.youtube.com/watch?v=I58HzUhQ3W8

In short, you'd draw the shadow iterations to a separate cache canvas, blur the result, then adjust the levels on the blurred result to eliminate the blurry regions. This evens out the shadow and would probably let you get away with fewer shadow iterations. Having a separate cache for the shadow would also allow you to control the shadow's opacity (not possible if you draw the shadow in the same cache as the text)
